modules/kafka/tests/test_kafka_topic.c

```diff
@@ -22,157 +22,181 @@
 }
 
-Test(kafka_topic, test_topics)
+ParameterizedTest(struct topic_validity_test_params *param, kafka_topic, topic_validity_tests)
 {
-  _init_topics();
-  cr_assert_str_eq(rd_kafka_topic_name(kafka_driver->topic), kafka_driver->topic_name->template);
-  cr_assert_str_eq(rd_kafka_topic_name(kafka_driver->fallback_topic), kafka_driver->fallback_topic_name);
+  KafkaDestDriver *driver = (KafkaDestDriver *) kafka_dd_new(configuration);
+  cr_assert_eq(kafka_dd_validate_topic_name(driver, param->topic_name), param->isvalid);
+  log_pipe_unref((LogPipe *)driver);
 }
 
-Test(kafka_topic, test_topic_name_validity)
+Test(kafka_topic, test_resolve_template_topic_name)
 {
-  cr_assert(_validate_topic_name(kafka_driver, "test 1") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, "") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, "..") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, ".") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, " invalidname") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, "aba#2") == FALSE);
-  cr_assert(_validate_topic_name(kafka_driver, g_strnfill(250, 'a')) == FALSE);
-
-  cr_assert(_validate_topic_name(kafka_driver, "validname123") == TRUE);
-  cr_assert(_validate_topic_name(kafka_driver, "valid.name123") == TRUE);
-  cr_assert(_validate_topic_name(kafka_driver, "valid_name123") == TRUE);
-  cr_assert(_validate_topic_name(kafka_driver, "valid-name123") == TRUE);
-  cr_assert(_validate_topic_name(kafka_driver, g_strnfill(249, 'a')) == TRUE);
-}
+  configuration = cfg_new_snippet();
+  LogDriver *driver = kafka_dd_new(configuration);
 
+  LogTemplate *topic_name = log_template_new(configuration, NULL);
+
+  log_template_compile(topic_name, "$MSG", NULL);
+  kafka_dd_set_topic(driver, topic_name);
+  kafka_dd_set_fallback_topic(driver, "fallbacktopicname");
+
+  g_assert(kafka_dd_init((LogPipe *) driver));
+
+  KafkaDestDriver *kafka_driver = (KafkaDestDriver *) driver;
+
+  KafkaDestWorker *worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
 
-Test(kafka_topic, test_valid_non_template_calculate_topic)
-{
-  _init_topics();
-  worker = (KafkaDestWorker *) kafka_dest_worker_new(&kafka_driver->super, 0);
   LogMessage *msg = log_msg_new_empty();
-  cr_assert_str_eq(rd_kafka_topic_name(_calculate_topic(worker, msg)), "topicname");
+  log_msg_set_value_by_name(msg, "MSG", "valid_template_topic", -1);
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "valid_template_topic");
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
+
+  log_msg_set_value_by_name(msg, "MSG", "invalid name", -1);
+
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "fallbacktopicname");
+  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
```

Comment:
I don't see the point of having both of these asserts (same for the other asserts in this function):
```
  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), "fallbacktopicname");
  cr_assert_str_eq(_resolve_template_topic_name(worker, msg), rd_kafka_topic_name(_calculate_topic(worker, msg)));
```
`_resolve_template_topic_name()` is called by `_calculate_topic()`, I don't understand the necessity of comparing them.
Which from these 2 functions do you want to test here?

Reply:
It was primarily to test both the functions. Should I write a separate test for just ```_calculate_topic()``` instead?

Reply:
That sounds good!